icu4c/source/i18n/units_converter.cpp

```diff
@@ -502,6 +502,45 @@ UnitConverter::UnitConverter(const MeasureUnitImpl &source, const MeasureUnitImp
                        ratesInfo, status);
 }
 
+int32_t UnitConverter::compareTwoUnits(const MeasureUnitImpl &firstUnit, const MeasureUnitImpl &secondUnit,
+                                       const ConversionRates &ratesInfo, UErrorCode &status) {
+    if (firstUnit.complexity == UMeasureUnitComplexity::UMEASURE_UNIT_MIXED ||
+        secondUnit.complexity == UMeasureUnitComplexity::UMEASURE_UNIT_MIXED) {
+        status = U_INTERNAL_PROGRAM_ERROR;
+        return 0;
+    }
+
+    Convertibility unitsState = extractConvertibility(firstUnit, secondUnit, ratesInfo, status);
+    if (U_FAILURE(status)) return 0;
+
+    if (unitsState == Convertibility::UNCONVERTIBLE || unitsState == Convertibility::RECIPROCAL) {
+        status = U_INTERNAL_PROGRAM_ERROR;
+        return 0;
+    }
+
+    // Represents the conversion factor from the firstUnit to the base unit that specified in the conversion
+    // data which is considered as the root of the firstUnit and the secondUnit.
+    Factor firstUnitToBase = loadCompoundFactor(firstUnit, ratesInfo, status);
+    Factor secondUnitToBase = loadCompoundFactor(secondUnit, ratesInfo, status);
+
+    firstUnitToBase.substituteConstants();
+    secondUnitToBase.substituteConstants();
+
+    double firstUnitToBaseConversionRate = firstUnitToBase.factorNum / firstUnitToBase.factorDen;
+    double secondUnitToBaseConversionRate = secondUnitToBase.factorNum / secondUnitToBase.factorDen;
+
+    double diff = firstUnitToBaseConversionRate - secondUnitToBaseConversionRate;
+    if (std::abs(diff) <= DBL_EPSILON) {
```

Comment:
This check doesn't do much. DBL_EPSILON is relevant relative to the value 1.0, while the conversion rates are probably far from 1.0. No unit tests fail when I delete this if. I think you can remove it.

Reply:
Rephrased: I do really like the intent of the logic, but it would probably require division, not subtraction, to check if the ratio is within epsilon of each other.

Considering division is relatively slow - and we probably don't care too much about sort stability for units considered to be the same, I think comparing diff to 0 makes sense.

(Consider two conversion rates that are both really tiny: DBL_EPSILON will start being a significant number.)

Reply:
E.g. attometer to meter conversion rate is 1e-18, that's smaller than DBL_EPSILON. So I suspect comparing attometer to attoinch and this comparison will claim they're the same - we could prove it with a unit test. :-)

Reply:
parsec to light-year, for example:

parsec = 3.086e+16 meters
factor1 -> 3.2404407e-17

light-year = 9.461e+15 meters
factor2 -> 1.0569707e-16

factor2 - factor1 = 1.0569707e-16 - 3.2404407e-17 = 7.3292663e-17
DBL_EPSILON = 2.2204460492503131e-16

However factor2/factor1 = 1.0569707e-16 / 3.2404407e-17 = 3.26181158013
That's what's relevant. (About 3.26 light-years to a parsec.)

So either compare the ratio to something like (1+DBL_EPSILON), or just don't worry about it and compare to zero.

Reply:
for this line
```
factor2 - factor1 = 1.0569707e-16 - 3.2404407e-17 = 7.3292663e-17
DBL_EPSILON = 2.2204460492503131e-16 // Bigger than factor2-factor1!
```

Do you think double can represent 7.3292663e-17 ?

by the way, I will just compare it to zero.

Reply:
Just FYI:

"Accurate to at least 15 digits", when wanting to simply express that constant value. But when doing this calculation, more precision is lost - that constant is itself actually already inaccurate.

Consider: using "bc", which can handle arbitrary precision, asking it to do 50 decimal places, and using the values for parsec and lightyear from units.txt:

```
scale=50
parsec=30856780000000000
1/parsec
.00000000000000003240778849899438632287620419240115
lightyear=9460730000000000
1/lightyear
.00000000000000010570008868237440451212538567319857
1/lightyear - 1/parsec
.00000000000000007329230018338001818924918148079742
```

If I'm doing this correctly, the more accurate answer should be nearer 7.329230018338e-17, not 7.3292663e-17 - so whatever I used to do the calculation before (Google?) already caused an error in the sixth decimal place. Hmmm.

Reply:
Ah, logical: above, I rounded parsec and light-year to 4 significant digits, and did my further calculations with that. The point was only to get answer accurate enough to show that the result is much smaller than DBL_EPSILON.